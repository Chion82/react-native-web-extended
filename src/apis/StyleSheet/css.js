const DISPLAY_FLEX_CLASSNAME = '__style_df';
const POINTER_EVENTS_AUTO_CLASSNAME = '__style_pea';
const POINTER_EVENTS_BOX_NONE_CLASSNAME = '__style_pebn';
const POINTER_EVENTS_BOX_ONLY_CLASSNAME = '__style_pebo';
const POINTER_EVENTS_NONE_CLASSNAME = '__style_pen';

/* eslint-disable max-len */
const CSS_RESET =
// reset unwanted styles
'/* React Native */\n' +
'html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0)}\n' +
'body{margin:0}\n' +
'button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}\n' +
'input::-webkit-inner-spin-button,input::-webkit-outer-spin-button,' +
'input::-webkit-search-cancel-button,input::-webkit-search-decoration,' +
'input::-webkit-search-results-button,input::-webkit-search-results-decoration{display:none}';

const CSS_HELPERS =
// vendor prefix 'display:flex' until React supports fallback values for inline styles
`.${DISPLAY_FLEX_CLASSNAME} {display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex}\n` +
// implement React Native's pointer event values
`.${POINTER_EVENTS_AUTO_CLASSNAME}, .${POINTER_EVENTS_BOX_ONLY_CLASSNAME}, .${POINTER_EVENTS_BOX_NONE_CLASSNAME} * {pointer-events:auto}\n` +
`.${POINTER_EVENTS_NONE_CLASSNAME}, .${POINTER_EVENTS_BOX_ONLY_CLASSNAME} *, .${POINTER_EVENTS_NONE_CLASSNAME} {pointer-events:none}`;
/* eslint-enable max-len */

const styleAsClassName = {
  display: {
    'flex': DISPLAY_FLEX_CLASSNAME
  },
  pointerEvents: {
    'auto': POINTER_EVENTS_AUTO_CLASSNAME,
    'box-none': POINTER_EVENTS_BOX_NONE_CLASSNAME,
    'box-only': POINTER_EVENTS_BOX_ONLY_CLASSNAME,
    'none': POINTER_EVENTS_NONE_CLASSNAME
  }
};

export const getDefaultStyleSheet = () => `${CSS_RESET}\n${CSS_HELPERS}`;

export const getStyleAsHelperClassName = (prop, value) => {
  return styleAsClassName[prop] && styleAsClassName[prop][value];
};
