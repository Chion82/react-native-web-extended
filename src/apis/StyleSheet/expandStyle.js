/**
 * The browser implements the CSS cascade, where the order of properties is a
 * factor in determining which styles to paint. React Native is different in
 * giving precedence to the more specific styles. For example, the value of
 * `paddingTop` takes precedence over that of `padding`.
 *
 * This module creates mutally exclusive style declarations by expanding all of
 * React Native's supported shortform properties (e.g. `padding`) to their
 * longfrom equivalents.
 */

import normalizeValue from './normalizeValue';

const emptyObject = {};
const styleShortFormProperties = {
  borderColor: [ 'borderTopColor', 'borderRightColor', 'borderBottomColor', 'borderLeftColor' ],
  borderRadius: [ 'borderTopLeftRadius', 'borderTopRightRadius', 'borderBottomRightRadius', 'borderBottomLeftRadius' ],
  borderStyle: [ 'borderTopStyle', 'borderRightStyle', 'borderBottomStyle', 'borderLeftStyle' ],
  borderWidth: [ 'borderTopWidth', 'borderRightWidth', 'borderBottomWidth', 'borderLeftWidth' ],
  margin: [ 'marginTop', 'marginRight', 'marginBottom', 'marginLeft' ],
  marginHorizontal: [ 'marginRight', 'marginLeft' ],
  marginVertical: [ 'marginTop', 'marginBottom' ],
  overflow: [ 'overflowX', 'overflowY' ],
  padding: [ 'paddingTop', 'paddingRight', 'paddingBottom', 'paddingLeft' ],
  paddingHorizontal: [ 'paddingRight', 'paddingLeft' ],
  paddingVertical: [ 'paddingTop', 'paddingBottom' ],
  textDecorationLine: [ 'textDecoration' ],
  writingDirection: [ 'direction' ]
};

const alphaSort = (arr) => arr.sort((a, b) => {
  if (a < b) { return -1; }
  if (a > b) { return 1; }
  return 0;
});

const createStyleReducer = (originalStyle) => {
  const originalStyleProps = Object.keys(originalStyle);

  return (style, prop) => {
    const value = normalizeValue(prop, originalStyle[prop]);
    const longFormProperties = styleShortFormProperties[prop];

    // React Native treats `flex:1` like `flex:1 1 auto`
    if (prop === 'flex') {
      style.flexGrow = value;
      style.flexShrink = 1;
      style.flexBasis = 'auto';
    // React Native accepts 'center' as a value
    } else if (prop === 'textAlignVertical') {
      style.verticalAlign = (value === 'center' ? 'middle' : value);
    } else if (longFormProperties) {
      longFormProperties.forEach((longForm, i) => {
        // the value of any longform property in the original styles takes
        // precedence over the shortform's value
        if (originalStyleProps.indexOf(longForm) === -1) {
          style[longForm] = value;
        }
      });
    } else {
      style[prop] = value;
    }
    return style;
  };
};

const expandStyle = (style = emptyObject) => {
  const sortedStyleProps = alphaSort(Object.keys(style));
  const styleReducer = createStyleReducer(style);
  return sortedStyleProps.reduce(styleReducer, {});
};

module.exports = expandStyle;
