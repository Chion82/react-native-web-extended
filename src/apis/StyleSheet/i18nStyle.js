import I18nManager from '../I18nManager';
import multiplyStyleLengthValue from '../../modules/multiplyStyleLengthValue';

const emptyObject = {};

/**
 * Map of property names to their BiDi equivalent.
 */
const PROPERTIES_TO_SWAP = {
  'borderTopLeftRadius': 'borderTopRightRadius',
  'borderTopRightRadius': 'borderTopLeftRadius',
  'borderBottomLeftRadius': 'borderBottomRightRadius',
  'borderBottomRightRadius': 'borderBottomLeftRadius',
  'borderLeftColor': 'borderRightColor',
  'borderLeftStyle': 'borderRightStyle',
  'borderLeftWidth': 'borderRightWidth',
  'borderRightColor': 'borderLeftColor',
  'borderRightWidth': 'borderLeftWidth',
  'borderRightStyle': 'borderLeftStyle',
  'left': 'right',
  'marginLeft': 'marginRight',
  'marginRight': 'marginLeft',
  'paddingLeft': 'paddingRight',
  'paddingRight': 'paddingLeft',
  'right': 'left'
};

const PROPERTIES_SWAP_LEFT_RIGHT = {
  'clear': true,
  'float': true,
  'textAlign': true
};

const PROPERTIES_SWAP_LTR_RTL = {
  'writingDirection': true
};

/**
 * Invert the sign of a numeric-like value
 */
const additiveInverse = (value: String | Number) => multiplyStyleLengthValue(value, -1);

/**
 * BiDi flip the given property.
 */
const flipProperty = (prop:String): String => {
  return PROPERTIES_TO_SWAP.hasOwnProperty(prop) ? PROPERTIES_TO_SWAP[prop] : prop;
};

/**
 * BiDi flip translateX
 */
const flipTransform = (transform: Object): Object => {
  const translateX = transform.translateX;
  if (translateX != null) {
    transform.translateX = additiveInverse(translateX);
  }
  return transform;
};

const swapLeftRight = (value:String): String => {
  return value === 'left' ? 'right' : value === 'right' ? 'left' : value;
};

const swapLtrRtl = (value:String): String => {
  return value === 'ltr' ? 'rtl' : value === 'rtl' ? 'ltr' : value;
};

const i18nStyle = (style = emptyObject) => {
  const newStyle = {};
  for (const prop in style) {
    if (!Object.prototype.hasOwnProperty.call(style, prop)) {
      continue;
    }

    const indexOfNoFlip = prop.indexOf('$noI18n');

    if (I18nManager.isRTL) {
      if (PROPERTIES_TO_SWAP[prop]) {
        const newProp = flipProperty(prop);
        newStyle[newProp] = style[prop];
      } else if (PROPERTIES_SWAP_LEFT_RIGHT[prop]) {
        newStyle[prop] = swapLeftRight(style[prop]);
      } else if (PROPERTIES_SWAP_LTR_RTL[prop]) {
        newStyle[prop] = swapLtrRtl(style[prop]);
      } else if (prop === 'textShadowOffset') {
        newStyle[prop] = style[prop];
        newStyle[prop].width = additiveInverse(style[prop].width);
      } else if (prop === 'transform') {
        newStyle[prop] = style[prop].map(flipTransform);
      } else if (indexOfNoFlip > -1) {
        const newProp = prop.substring(0, indexOfNoFlip);
        newStyle[newProp] = style[prop];
      } else {
        newStyle[prop] = style[prop];
      }
    } else {
      if (indexOfNoFlip > -1) {
        const newProp = prop.substring(0, indexOfNoFlip);
        newStyle[newProp] = style[prop];
      } else {
        newStyle[prop] = style[prop];
      }
    }
  }

  return newStyle;
};

module.exports = i18nStyle;
