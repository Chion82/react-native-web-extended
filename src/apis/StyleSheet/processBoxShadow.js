import normalizeColor from '../../modules/normalizeColor';
import normalizeValue from './normalizeValue';

const applyOpacity = (color, opacity) => {
  const normalizedColor = normalizeColor(color);
  const colorNumber = normalizedColor === null ? 0x00000000 : normalizedColor;
  const r = (colorNumber & 0xff000000) >>> 24;
  const g = (colorNumber & 0x00ff0000) >>> 16;
  const b = (colorNumber & 0x0000ff00) >>> 8;
  const a = (((colorNumber & 0x000000ff) >>> 0) / 255).toFixed(2);
  return `rgba(${r},${g},${b},${a * opacity})`;
};

// TODO: add inset and spread support
const processBoxShadow = (style) => {
  if (style && style.shadowColor) {
    const { height, width } = style.shadowOffset || {};
    const opacity = style.shadowOpacity != null ? style.shadowOpacity : 1;
    const color = applyOpacity(style.shadowColor, opacity);
    const blurRadius = normalizeValue(null, style.shadowRadius || 0);
    const offsetX = normalizeValue(null, height || 0);
    const offsetY = normalizeValue(null, width || 0);
    const boxShadow = `${offsetX} ${offsetY} ${blurRadius} ${color}`;
    style.boxShadow = style.boxShadow ? `${style.boxShadow}, ${boxShadow}` : boxShadow;
  }
  delete style.shadowColor;
  delete style.shadowOffset;
  delete style.shadowOpacity;
  delete style.shadowRadius;
  return style;
};

module.exports = processBoxShadow;
