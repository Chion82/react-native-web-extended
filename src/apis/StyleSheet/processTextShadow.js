import normalizeValue from './normalizeValue';

const processTextShadow = (style) => {
  if (style && style.textShadowOffset) {
    const { height, width } = style.textShadowOffset;
    const offsetX = normalizeValue(null, height || 0);
    const offsetY = normalizeValue(null, width || 0);
    const blurRadius = normalizeValue(null, style.textShadowRadius || 0);
    const color = style.textShadowColor || 'currentcolor';

    style.textShadow = `${offsetX} ${offsetY} ${blurRadius} ${color}`;
    style.textShadowColor = null;
    style.textShadowOffset = null;
    style.textShadowRadius = null;
  }
  return style;
};

module.exports = processTextShadow;
