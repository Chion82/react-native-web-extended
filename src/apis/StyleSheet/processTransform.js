import normalizeValue from './normalizeValue';

// { scale: 2 } => 'scale(2)'
// { translateX: 20 } => 'translateX(20px)'
const mapTransform = (transform) => {
  const type = Object.keys(transform)[0];
  const value = normalizeValue(type, transform[type]);
  return `${type}(${value})`;
};

// [1,2,3,4,5,6] => 'matrix3d(1,2,3,4,5,6)'
const convertTransformMatrix = (transformMatrix) => {
  const matrix = transformMatrix.join(',');
  return `matrix3d(${matrix})`;
};

const processTransform = (style) => {
  if (style) {
    if (style.transform && Array.isArray(style.transform)) {
      style.transform = style.transform.map(mapTransform).join(' ');
    } else if (style.transformMatrix) {
      style.transform = convertTransformMatrix(style.transformMatrix);
      style.transformMatrix = null;
    }
  }
  return style;
};

module.exports = processTransform;
