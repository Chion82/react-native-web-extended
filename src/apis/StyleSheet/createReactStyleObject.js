import expandStyle from './expandStyle';
import flattenStyle from '../../modules/flattenStyle';
import i18nStyle from './i18nStyle';
import processBoxShadow from './processBoxShadow';
import processTextShadow from './processTextShadow';
import processTransform from './processTransform';
import processVendorPrefixes from './processVendorPrefixes';

const processors = [
  processBoxShadow,
  processTextShadow,
  processTransform,
  processVendorPrefixes
];

const applyProcessors = (style) => processors.reduce((style, processor) => processor(style), style);

const createReactDOMStyleObject = (reactNativeStyle) => applyProcessors(
  expandStyle(i18nStyle(flattenStyle(reactNativeStyle)))
);

module.exports = createReactDOMStyleObject;
