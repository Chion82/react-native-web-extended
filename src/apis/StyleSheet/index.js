import * as css from './css';
import createReactStyleObject from './createReactStyleObject';
import ExecutionEnvironment from 'fbjs/lib/ExecutionEnvironment';
import flattenStyle from '../../modules/flattenStyle';
import React from 'react';
import ReactNativePropRegistry from '../../modules/ReactNativePropRegistry';

let styleElement;
let shouldInsertStyleSheet = ExecutionEnvironment.canUseDOM;

const STYLE_SHEET_ID = 'react-native-style__';

const absoluteFillObject = { position: 'absolute', left: 0, right: 0, top: 0, bottom: 0 };

const defaultStyleSheet = css.getDefaultStyleSheet();

const insertStyleSheet = () => {
  // check if the server rendered the style sheet
  styleElement = document.getElementById(STYLE_SHEET_ID);
  // if not, inject the style sheet
  if (!styleElement) {
    document.head.insertAdjacentHTML(
      'afterbegin',
      `<style id="${STYLE_SHEET_ID}">${defaultStyleSheet}</style>`
    );
    shouldInsertStyleSheet = false;
  }
};

module.exports = {
  /**
   * For testing
   * @private
   */
  _reset() {
    if (styleElement) {
      document.head.removeChild(styleElement);
      styleElement = null;
      shouldInsertStyleSheet = true;
    }
  },

  absoluteFill: ReactNativePropRegistry.register(absoluteFillObject),

  absoluteFillObject,

  create(styles) {
    if (shouldInsertStyleSheet) {
      insertStyleSheet();
    }

    const result = {};
    for (const key in styles) {
      if (process.env.NODE_ENV !== 'production') {
        require('./StyleSheetValidation').validateStyle(key, styles);
      }
      result[key] = ReactNativePropRegistry.register(styles[key]);
    }
    return result;
  },

  hairlineWidth: 1,

  flatten: flattenStyle,

  /* @platform web */
  render() {
    return <style dangerouslySetInnerHTML={{ __html: defaultStyleSheet }} id={STYLE_SHEET_ID} />;
  },

  /**
   * Accepts React props and converts style declarations to classNames when necessary
   * @platform web
   */
  resolve(props) {
    let className = props.className || '';
    const style = createReactStyleObject(props.style);
    for (const prop in style) {
      const value = style[prop];
      const replacementClassName = css.getStyleAsHelperClassName(prop, value);
      if (replacementClassName) {
        className += ` ${replacementClassName}`;
        style[prop] = null;
      }
    }

    return { className, style };
  }
};
