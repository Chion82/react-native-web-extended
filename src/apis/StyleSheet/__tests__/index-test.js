/* eslint-env jasmine, jest */

import { getDefaultStyleSheet } from '../css';
import StyleSheet from '..';

const isPlainObject = (x) => {
  const toString = Object.prototype.toString;
  let proto;
  /* eslint-disable */
  return (
    toString.call(x) === '[object Object]' &&
    (proto = Object.getPrototypeOf(x), proto === null || proto === Object.getPrototypeOf({}))
  );
  /* eslint-enable */
};

describe('apis/StyleSheet', () => {
  beforeEach(() => {
    StyleSheet._reset();
  });

  test('absoluteFill', () => {
    expect(Number.isInteger(StyleSheet.absoluteFill) === true).toBeTruthy();
  });

  test('absoluteFillObject', () => {
    expect(isPlainObject(StyleSheet.absoluteFillObject) === true).toBeTruthy();
  });

  describe('create', () => {
    test('replaces styles with numbers', () => {
      const style = StyleSheet.create({ root: { opacity: 1 } });
      expect(Number.isInteger(style.root) === true).toBeTruthy();
    });

    test('renders a style sheet in the browser', () => {
      StyleSheet.create({ root: { color: 'red' } });
      expect(document.getElementById('react-native-style__').textContent).toEqual(getDefaultStyleSheet());
    });
  });

  test('flatten', () => {
    expect(typeof StyleSheet.flatten === 'function').toBeTruthy();
  });

  test('hairlineWidth', () => {
    expect(Number.isInteger(StyleSheet.hairlineWidth) === true).toBeTruthy();
  });

  test('render', () => {
    expect(StyleSheet.render().props.dangerouslySetInnerHTML.__html).toEqual(getDefaultStyleSheet());
  });

  test('resolve', () => {
    expect(StyleSheet.resolve({
      className: 'test',
      style: {
        display: 'flex',
        opacity: 1,
        pointerEvents: 'box-none'
      }
    })).toMatchSnapshot();
  });
});
