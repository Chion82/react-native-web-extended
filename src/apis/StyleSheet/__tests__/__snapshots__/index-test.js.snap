exports[`apis/StyleSheet resolve 1`] = `
Object {
  "className": "test __style_df __style_pebn",
  "style": Object {
    "display": null,
    "opacity": 1,
    "pointerEvents": null,
  },
}
`;
